src/core/job.c

```diff
@@ -878,7 +897,17 @@ int job_finish_and_invalidate(Job *j, JobResult result, bool recursive, bool alr
                 j->manager->n_failed_jobs++;
 
         job_uninstall(j);
-        job_free(j);
+        /* Remember jobs started before the reload */
+        if (MANAGER_IS_RELOADING(j->manager) && j->reloaded) {
+            int r;
+
+            job_unlink(j);
+            r = set_put(j->manager->pending_finished_jobs, j);
+            if (r < 0)
```

Comment:
Since you don't need the value of `r` anywhere, just collapse this into
```C
if (set_put(j->manager->pending_finished_jobs, j) < 0)
    job_free(j);
```